### PR TITLE
Misc: cleanup, fixes, improvements, docs

### PR DESCRIPTION
**Pull request by FabioLuporini**
Yet another consolidation PR

- Improved (up-to-date) docs
- Fix output reports (reported iteration counts might be a small quantity (O(1)) off)
- Refactor the ``codepy.Compiler`` hierarchy and (finally!) get rid of all annoying gcc-induced warnings. Travis' trace will now be much cleaner. 
- If the specific compiler version supports OpenMP 4.0, we now unconditionally add the appropriate ``-openmp`` flag so that ``#pragma simd omp`` are reconognised and vectorization properly triggered
- Drop support for Intel MIC (untested => broken; unused => useless; broken + useless => drop)
- Fix a subtle bug by adding a specialised ``_hashable_content`` to ``Dimensions``
- Minor fixes in YASK

### CODE REVIEW

File: devito/function.py [hunk 1]
```diff
@@ -47,6 +47,12 @@ class Constant(AbstractSymbol, ConstantArgProvider):
 
     """
     Symbol representing constant values in symbolic equations.
+
+    .. note::
+
+        The parameters must always be given as keyword arguments, since
+        SymPy uses `*args` to (re-)create the dimension arguments of the
```
Comment (mlange05) [inline devito/function.py]: If memory serves, double ` is needed to produce correct teletype in the docs. Also below...

Comment (FabioLuporini) [reply devito/function.py]: fixed

**Comment by FabioLuporini** [comment]
OK, I also pushed a couple of fixes that were making Travis rightfully angry. Will merge once Travis is green.

**Comment by FabioLuporini** [comment]
OK, merging...

**Comment by FabioLuporini** [comment]
Merged.

**Pull request merged**